### PR TITLE
[feature] Support datasets with data defined on faces in mesh calculator

### LINKED ISSUE #30170: Mesh calculator is sluggish and unresponsive
When I try to type something in the mesh calculator expression box, there is a lag between when I type and the time, it appears on the screen. It seems validation of the expression is taking a lot of resources in the background.

### LINKED ISSUE #30219: Support for cells in mesh calculator
Currently, mesh calculator only supports vertex-centred meshes. But in some cases (e.g. UGRID netcdf), values are stored on the edges.

### PR DESCRIPTION
**Pull request by PeterPetrik**
fix #30219
fix #30170 

added "driver" and "group name" to the calculator interface. MDAL now supports 3 drivers for storing results, so user must be able to choose appropriate driver and dataset group name (some drivers store multiple groups to 1 file)

Screenshot 2019-12-05 at 16 14 00

### CODE REVIEW

File: python/analysis/auto_generated/mesh/qgsmeshcalculator.sip.in [hunk 1]
```diff
@@ -46,6 +45,8 @@ implemented and supported
     };
 
     QgsMeshCalculator( const QString &formulaString,
+                       const QString &outputDriver,
```
Comment (nyalldawson) [inline python/analysis/auto_generated/mesh/qgsmeshcalculator.sip.in]: This is an API break, right?

File: python/analysis/auto_generated/mesh/qgsmeshcalculator.sip.in [hunk 2]
```diff
@@ -96,6 +103,19 @@ Returns whether formula is valid for particular mesh layer
 :param layer: mesh layer with dataset groups references in formulaString
 
 :return: QgsMeshCalculator.Success in case of success
+%End
+
+    static Result expression_valid( const QString &formulaString, QgsMeshLayer *layer, QgsMeshDriverMetadata::MeshDriverCapability &requiredCapability );
```
Comment (nyalldawson) [inline python/analysis/auto_generated/mesh/qgsmeshcalculator.sip.in]: isExpressionValid? expressionIsValid?

File: src/analysis/mesh/qgsmeshcalculator.cpp [hunk 3]
```diff
@@ -179,7 +194,7 @@ QgsMeshCalculator::Result QgsMeshCalculator::processCalculation( QgsFeedback *fe
 
   const QgsMeshDatasetGroupMetadata meta = outputGroup->groupMetadata();
   bool err = mMeshLayer->dataProvider()->persistDatasetGroup(
-               mOutputFile,
+               mOutputDriver + ":" + mOutputFile,
```
Comment (nyalldawson) [inline src/analysis/mesh/qgsmeshcalculator.cpp]: ':'

File: src/analysis/mesh/qgsmeshcalcutils.cpp [hunk 4]
```diff
@@ -57,10 +64,78 @@ std::shared_ptr<QgsMeshMemoryDatasetGroup> QgsMeshCalcUtils::create( const QStri
...
+              Q_ASSERT( data.size() == resultCount );
+              for ( int value_i = 0; value_i < resultCount; ++value_i )
+                ds->values[value_i] = QgsMeshDatasetValue( data[value_i] );
+            }
+            else
+            {
+              QVector<double> buf = block.values();
+              QVector<double> x( nativeCount );
+              QVector<double> y( nativeCount );
+              for ( int value_i = 0; value_i < nativeCount; ++value_i )
```
Comment (nyalldawson) [inline src/analysis/mesh/qgsmeshcalcutils.cpp]: Naming convention

File: src/analysis/mesh/qgsmeshcalcutils.cpp [hunk 5]
```diff
@@ -85,9 +160,10 @@ std::shared_ptr<QgsMeshMemoryDataset> QgsMeshCalcUtils::create( const QgsMeshMem
 
 std::shared_ptr<QgsMeshMemoryDataset> QgsMeshCalcUtils::create( const QgsMeshDatasetGroupMetadata::DataType type ) const
 {
+  assert( type != QgsMeshDatasetGroupMetadata::DataOnVolumes );
```
Comment (nyalldawson) [inline src/analysis/mesh/qgsmeshcalcutils.cpp]: Q_ASSERT

File: src/analysis/mesh/qgsmeshcalcutils.cpp [hunk 6]
```diff
@@ -219,7 +299,7 @@ void QgsMeshCalcUtils::populateSpatialFilter( QgsMeshMemoryDatasetGroup &filter,
   const QList<int> faceIndexesForRectangle = triangularMesh()->faceIndexesForRectangle( extent );
   const QVector<int> trianglesToNativeFaces = triangularMesh()->trianglesToNativeFaces();
 
-  if ( mOutputType == QgsMeshDatasetGroupMetadata::DataOnVertices )
+  if ( mOutputType  == QgsMeshDatasetGroupMetadata::DataOnVertices )
```
Comment (nyalldawson) [inline src/analysis/mesh/qgsmeshcalcutils.cpp]: Remove extra space

File: src/analysis/mesh/qgsmeshcalcutils.cpp [hunk 7]
```diff
@@ -1079,16 +1166,41 @@ void QgsMeshCalcUtils::maximum( QgsMeshMemoryDatasetGroup &group1, const QgsMesh
   return func2( group1, group2, std::bind( & QgsMeshCalcUtils::fmax, this, std::placeholders::_1, std::placeholders::_2 ) );
 }
 
+QgsMeshDatasetGroupMetadata::DataType QgsMeshCalcUtils::determineResultDataType( QgsMeshLayer *layer, const QStringList &usedGroupNames )
+{
+  QHash<QString, int> names;
+  const auto dp = layer->dataProvider();
```
Comment (nyalldawson) [inline src/analysis/mesh/qgsmeshcalcutils.cpp]: Incorrect use of auto, use explicit types here

File: src/analysis/mesh/qgsmeshcalcutils.h [hunk 8]
```diff
@@ -54,6 +54,9 @@ class ANALYSIS_EXPORT QgsMeshCalcUtils
 
     /**
      * Creates the utils and validates the input
+     *
+     * The contructor fetches all dataset values and creates memory datasets from them.
```
Comment (nyalldawson) [inline src/analysis/mesh/qgsmeshcalcutils.h]: Constructor

File: src/app/mesh/qgsmeshcalculatordialog.cpp [hunk 9]
```diff
@@ -215,6 +235,58 @@ void QgsMeshCalculatorDialog::toggleExtendMask( int state )
...
+  }
+  else
+  {
+    mButtonBox->button( QDialogButtonBox::Ok )->setEnabled( false );
+    if ( !expressionValid )
+      mExpressionValidLabel->setText( tr( "Expression invalid" ) );
+    else if ( !filePathValid )
+      mExpressionValidLabel->setText( tr( "Invalid file path" ) );
+    else if ( !driverValid )
+      mExpressionValidLabel->setText( tr( "Required driver that can store data defined on %1" ).arg( requiredCapability == QgsMeshDriverMetadata::CanWriteFaceDatasets ? tr( " faces " ) : tr( " vertices " ) ) );
```
Comment (nyalldawson) [inline src/app/mesh/qgsmeshcalculatordialog.cpp]: "selected driver cannot store data defined on ..."

File: src/app/mesh/qgsmeshcalculatordialog.cpp [hunk 10]
```diff
@@ -442,6 +469,35 @@ QString QgsMeshCalculatorDialog::addSuffix( const QString fileName ) const
   return fileName + allowedSuffix;
 }
 
+void QgsMeshCalculatorDialog::getMeshDrivers()
+{
+  QgsProviderMetadata *providerMetadata = QgsProviderRegistry::instance()->providerMetadata( QStringLiteral( "mdal" ) );
+  if ( providerMetadata )
+  {
+    const auto allDrivers = providerMetadata->meshDriversMetadata();
```
Comment (nyalldawson) [inline src/app/mesh/qgsmeshcalculatordialog.cpp]: Incorrect auto use

File: src/providers/mdal/qgsmdalprovider.cpp [hunk 11]
```diff
@@ -190,7 +190,21 @@ bool QgsMdalProvider::persistDatasetGroup( const QString &path,
...
 
-  const QString driverName( "BINARY_DAT" ); //nothing else is implemented in MDAL
+  if ( path.isEmpty() )
+    return true;
+
+  // Form DRIVER:filename
+  QString filename = path;
+  // ASCII dat supports both face and vertex datasets
+  QString driverName = QStringLiteral( "DAT" );
+  QStringList parts = path.split( ":" );
```
Comment (nyalldawson) [inline src/providers/mdal/qgsmdalprovider.cpp]: ':'

File: src/providers/mdal/qgsmdalprovider.cpp [hunk 12]
```diff
@@ -190,7 +190,21 @@ bool QgsMdalProvider::persistDatasetGroup( const QString &path,
...
+  if ( parts.size() > 1 )
+  {
+    driverName = parts[0];
+    parts.removeFirst();
+    filename = parts.join( QStringLiteral( "" ) );
```
Comment (nyalldawson) [inline src/providers/mdal/qgsmdalprovider.cpp]: QStringLiteral("") -> QString()

File: python/analysis/auto_generated/mesh/qgsmeshcalculator.sip.in (hunk 1 above)
Comment (PeterPetrik) [reply python/analysis/auto_generated/mesh/qgsmeshcalculator.sip.in]: I could teoretically encode it all to previous "outputFile" parameter is some form list DRIVER:path:groupname or so, but not sure if it is worth to strictly keep the mesh API (where it has all over the place notice it is EXPERIMENTAL)...

File: src/core/mesh/qgsmeshdataprovider.h [hunk 13]
```diff
@@ -210,6 +210,11 @@ class CORE_EXPORT QgsMeshDataBlock
      */
     void *buffer() SIP_SKIP;
 
+    /**
+     * Returns buffer to the array with values
```
Comment (wonder-sk) [inline src/core/mesh/qgsmeshdataprovider.h]: worth adding a note it was added in 3.12

File: python/analysis/auto_generated/mesh/qgsmeshcalculator.sip.in (hunk 1 above)
Comment (PeterPetrik) [reply python/analysis/auto_generated/mesh/qgsmeshcalculator.sip.in]: added that api back with deprecated flag

**Review by wonder-sk — APPROVED** [review]

**Pull request merged**